benchmarks/instruction_counts/applications/ci.py

```diff
@@ -0,0 +1,84 @@
+"""Collect instruction counts for continuous integration."""
+import argparse
+import hashlib
+import json
+import time
+from typing import Dict, List, Optional, Union
+import uuid
+
+from core.expand import materialize
+from definitions.standard import BENCHMARKS
+from execution.runner import Runner
+from execution.work import WorkOrder
+
+
+REPEATS = 5
+TIMEOUT = 600  # Seconds
+RETRIES = 2
+
+VERSION = 0
+MD5 = "b984f1ebad546ed3ee612244b8c72e53"
+
+
+def main(argv: List[str]) -> None:
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--destination", type=str, default=None)
+    parser.add_argument("--subset", action="store_true")
+    args = parser.parse_args(argv)
+
+    t0 = int(time.time())
+    version = VERSION
+    benchmarks = materialize(BENCHMARKS)
+
+    # Useful for local development, since e2e time for the full suite is O(1 hour)
+    in_debug_mode = (args.subset or args.destination is None)
+    if args.subset:
+        version = -1
+        benchmarks = benchmarks[:10]
+
+    work_orders = tuple(
+        WorkOrder(label, autolabels, timer_args, timeout=TIMEOUT, retries=RETRIES)
+        for label, autolabels, timer_args in benchmarks * REPEATS
+    )
+
+    keys = tuple({str(work_order): None for work_order in work_orders}.keys())
+    md5 = hashlib.md5()
+    for key in keys:
+        md5.update(key.encode("utf-8"))
+
+    # Fail early, since collection takes a long time.
+    if md5.hexdigest() != MD5 and not args.subset:
+        msg = f"Expected {MD5}, got {md5.hexdigest()} instead"
+        if in_debug_mode:
+            print(f"WARNING: {msg}")
+        else:
+            raise ValueError(msg)
+
+    results = Runner(work_orders).run()
+
+    # TODO: Annotate with TypedDict when 3.8 is the minimum supported verson.
+    grouped_results: Dict[str, Dict[str, List[Union[float, int]]]] = {
+        key: {"times": [], "counts": []} for key in keys}
+
+    for work_order, r in results.items():
+        key = str(work_order)
+        grouped_results[key]["times"].extend(r.wall_times)
+        grouped_results[key]["counts"].extend(r.instructions)
+
+    final_results = {
+        "run_id": str(uuid.uuid4()),
+        "version": version,
+        "md5": md5.hexdigest(),
+        "start_time": t0,
+        "end_time": int(time.time()),
+        "values": grouped_results,
+    }
+
+    if args.destination is None:
```

Comment:
If this is for local debug, should we use `in_debug_mode` instead?

Reply:
It's a fair point. I split the file writing to one conditional, and put the print and pdb in a separate one conditioned on `in_debug_mode`. (Which I think is cleaner.)